node/subsystem-util/src/validator_discovery.rs

```diff
@@ -0,0 +1,155 @@
+
+/// Error when making a request to connect to validators. 
+#[derive(Debug, derive_more::From)]
+pub enum Error {
+	/// Attempted to send or receive on a oneshot channel which had been canceled
+	#[from]
+	Oneshot(oneshot::Canceled),
+	/// A subsystem error.
+	#[from]
+	Subsystem(SubsystemError),
+	/// An error in the Runtime API.
+	#[from]
+	RuntimeApi(RuntimeApiError),
+}
+
+/// Utility function to make it easier to connect to validators.
+pub async fn connect_to_validators<Context: SubsystemContext>(
+	ctx: &mut Context,
+	relay_parent: Hash,
+	validators: Vec<ValidatorId>,
+) -> Result<ConnectionRequest, Error> {
+	// ValidatorId -> AuthorityDiscoveryId
+	let (tx, rx) = oneshot::channel();
+
+	ctx.send_message(AllMessages::RuntimeApi(
+		RuntimeApiMessage::Request(
+			relay_parent,
+			RuntimeApiRequest::ValidatorDiscovery(validators.clone(), tx),
+		)
+	)).await?;
+
+	let maybe_authorities = rx.await??;
+	let authorities: Vec<_> = maybe_authorities.iter()
+		.cloned()
+		.filter_map(|id| id)
+		.collect();
+
+	let validator_map = validators.into_iter()
+		.zip(maybe_authorities.into_iter())
+		.filter_map(|(k, v)| v.map(|v| (v, k)))
+		.collect::<HashMap<AuthorityDiscoveryId, ValidatorId>>();
+
+	let (connections, revoke) = connect_to_authorities(ctx, authorities).await?;
+
+	Ok(ConnectionRequest {
+		validator_map,
+		connections,
+		revoke,
+	})
+}
+
+async fn connect_to_authorities<Context: SubsystemContext>(
+	ctx: &mut Context,
+	validator_ids: Vec<AuthorityDiscoveryId>,
+) -> Result<(mpsc::Receiver<(AuthorityDiscoveryId, PeerId)>, oneshot::Sender<()>), Error> {
+	const PEERS_CAPACITY: usize = 8;
+
+	let (revoke_tx, revoke) = oneshot::channel();
+	let (connected, connected_rx) = mpsc::channel(PEERS_CAPACITY);
+
+	ctx.send_message(AllMessages::NetworkBridge(
+		NetworkBridgeMessage::ConnectToValidators {
+			validator_ids,
+			connected,
+			revoke,
+		}
+	)).await?;
+
+	Ok((connected_rx, revoke_tx))
+}
+
+/// A pending connection request to validators.
+/// This struct implements `Stream` to allow for asynchronous
+/// discovery of validator addresses.
+///
+/// NOTE: you should call `revoke` on this struct
+/// when you're no longer interested in the requested validators.
+#[pin_project]
+pub struct ConnectionRequest {
+	#[pin]
+	validator_map: HashMap<AuthorityDiscoveryId, ValidatorId>,
+	#[pin]
+	#[must_use = "streams do nothing unless polled"]
+	connections: mpsc::Receiver<(AuthorityDiscoveryId, PeerId)>,
+	#[must_use = "a request should be revoked at some point"]
+	revoke: oneshot::Sender<()>,
+}
+
+impl stream::Stream for ConnectionRequest {
+	type Item = (ValidatorId, PeerId);
+
+	fn poll_next(self: Pin<&mut Self>, cx: &mut task::Context) -> Poll<Option<Self::Item>> {
+		let mut this = self.project();
+		if this.validator_map.is_empty() {
+			return Poll::Ready(None);
+		}
+		match this.connections.poll_next(cx) {
+			Poll::Ready(Some((id, peer_id))) => {
+				if let Some(validator_id) = this.validator_map.remove(&id) {
```

Comment:
```suggestion
		if self.validator_map.is_empty() {
			return Poll::Ready(None);
		}
		match Pin::new(&mut self.connections).poll_next(cx) {
			Poll::Ready(Some((id, peer_id))) => {
				if let Some(validator_id) = self.validator_map.remove(&id) {
```

Reply:
Doesn't seem to work
https://gitlab.parity.io/parity/polkadot/-/jobs/659694#L586